### ISSUE TITLE
Use TypedFunc instead of Func for wasmi functions

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#3820]
https://github.com/project-oak/oak/blob/b349d2dec0b0ae51a2eab1fb7a50aa20254c5b18/oak_functions/wasm/src/lib.rs#L391-L406

@mariaschett would it be possible to use TypedFunc instead of Func? As suggested in https://docs.rs/wasmi/0.29.0/wasmi/struct.TypedFunc.html#note .

**Comment by mariaschett** [comment]
We can only get a `Func` with [`into_func`](https://docs.rs/wasmi/0.29.0/wasmi/struct.Export.html#method.into_func) at this point. We could then convert to [typed](https://docs.rs/wasmi/0.29.0/wasmi/struct.Func.html#method.typed).

But I don't think that gives us the benefit we are after, as we'd still do that with every call to `get_alloc`.

We could store `typed_alloc` in the `OakCaller`, though. Do you think this is worth it?

**Comment by tiziano88** [comment]
Yes that seems exactly the use case: create the TypedFunc once from the Func and store it somewhere, so that future invocations don't need to perform all the type checks again.
